### PR TITLE
Added support for ca certificate file setting in SSL_CTX.

### PR DESCRIPTION
**Pull request by jbevemyr**
## Description

Added support for custom ca certificate roots.

## Testing

New tests are needed but I'm not familiar with the test setup yet.

## Documentation

The new CaCertificateFile option needs to be documented.

### CODE REVIEW

**Comment by jbevemyr** [comment]
@microsoft-github-policy-service agree

File: src/platform/tls_openssl.c [hunk 1]
```diff
@@ -1462,6 +1462,21 @@ CxPlatTlsSecConfigCreate(
         }
     }
 
+    if (CredConfig->CaCertificateFile) {
```
Comment (nibanks) [inline src/platform/tls_openssl.c:1465]: On Schannel, since this isn't supported, we should return an error if this is specified (not_supported or invalid_parameter).

File: src/inc/msquic.h [hunk 2]
```diff
@@ -119,6 +119,7 @@ typedef enum QUIC_CREDENTIAL_FLAGS {
     QUIC_CREDENTIAL_FLAG_CACHE_ONLY_URL_RETRIEVAL               = 0x00020000, // Windows only currently
     QUIC_CREDENTIAL_FLAG_REVOCATION_CHECK_CACHE_ONLY            = 0x00040000, // Windows only currently
     QUIC_CREDENTIAL_FLAG_INPROC_PEER_CERTIFICATE                = 0x00080000, // Schannel only
+    QUIC_CREDENTIAL_FLAG_USE_CACERTFILE                         = 0x00100000, // OpenSSL only currently
```
Comment (nibanks) [inline src/inc/msquic.h:122]: Let's call it `QUIC_CREDENTIAL_FLAG_SET_CA_CERTIFICATE_FILE` instead.

File: src/platform/tls_openssl.c [hunk 3]
```diff
@@ -1462,6 +1462,30 @@ CxPlatTlsSecConfigCreate(
...
+                    ERR_get_error(),
+                    "SSL_CTX_load_verify_locations failed");
+                Status = QUIC_STATUS_TLS_ERROR;
+                goto Exit;
+            }
+        } else {
+            QuicTraceEvent(
+                LibraryErrorStatus,
+                "QUIC_CREDENTIAL_FLAG_USE_CACERTFILE is set but "
+                "CaCertificateFile is not set.");
```
Comment (nibanks) [inline src/platform/tls_openssl.c:1483]: I don't think we need to error out if the flag is set but the pointer is NULL. Let's just treat the flag as "please check the pointer". So please combine the two top-level `if` checks into one, and only do anything if both are true.

File: src/platform/tls_openssl.c [hunk 4]
```diff
@@ -1462,6 +1462,30 @@ CxPlatTlsSecConfigCreate(
         }
     }
 
+    if (CredConfigFlags & QUIC_CREDENTIAL_FLAG_USE_CACERTFILE) {
+        if (CredConfig->CaCertificateFile) {
+            Ret = SSL_CTX_load_verify_locations(SecurityConfig->SSLCtx,
```
Comment (nibanks) [inline src/platform/tls_openssl.c:1467]: Would you mind updating the formatting to match more like the call to `SSL_CTX_use_certificate` above?

Comment (jbevemyr) [reply src/platform/tls_openssl.c:1467]: Sorry, I've been working in another repo with wierd indentation lately :-)

File: src/platform/tls_schannel.c [hunk 5]
```diff
@@ -1108,6 +1108,10 @@ CxPlatTlsSecConfigCreate(
     if (CredConfig->Flags & QUIC_CREDENTIAL_FLAG_REVOCATION_CHECK_CACHE_ONLY) {
         Credentials->dwFlags |= SCH_CRED_REVOCATION_CHECK_CACHE_ONLY;
     }
+    if (CredConfig->Flags & QUIC_CREDENTIAL_FLAG_SET_CA_CERTIFICATE_FILE) {
+        Status = QUIC_STATUS_NOT_SUPPORTED;
+        goto Error;
+    }
```
Comment (nibanks) [inline src/platform/tls_schannel.c:1114]: Please move this to the top of the function where all the other error-out conditions are?

**Comment by nibanks** [comment]
Everything is looking good. We just need some tests please.

**Comment by jbevemyr** [comment]
I'll try to add some tests.

File: src/platform/tls_openssl.c [hunk 6]
```diff
@@ -1644,6 +1644,7 @@ CxPlatTlsInitialize(
     uint16_t ServerNameLength = 0;
     UNREFERENCED_PARAMETER(State);
 
+
```
Comment (nibanks) [inline src/platform/tls_openssl.c:1647]: nit: remove extra newline?

File: src/inc/msquic.h [hunk 7]
```diff
@@ -317,7 +317,7 @@ typedef struct QUIC_CREDENTIAL_CONFIG {
     void* Reserved; // Currently unused
     QUIC_CREDENTIAL_LOAD_COMPLETE_HANDLER AsyncHandler; // Optional
     QUIC_ALLOWED_CIPHER_SUITE_FLAGS AllowedCipherSuites;// Optional
-    const char* CaCertificateFile;
+    char* CaCertificateFile;
```
Comment (nibanks) [inline src/inc/msquic.h:320]: Why can't it be const? It should be.

Comment (jbevemyr) [reply src/inc/msquic.h]: Yes, right.

File: src/platform/selfsign_openssl.c [hunk 8]
```diff
@@ -85,27 +86,27 @@ static uint8_t* ReadPkcs12(const char* Name, uint32_t* Length) {
...
+
+
 QUIC_STATUS
-CxPlatTlsGenerateSelfSignedCert(
-    _In_z_ char *CertFileName,
-    _When_(OutputPkcs12 == TRUE, _Reserved_)
-    _When_(OutputPkcs12 == FALSE, _In_z_)
-        char *PrivateKeyFileName,
+CxPlatTlsGenerateSignedCert(
+    _In_opt_z_ X509 *CaX509,
```
Comment (anrossi) [inline src/platform/selfsign_openssl.c:95]: This is not a null-terminated byte buffer, so you'd use `_In_opt_` instead.

File: src/platform/selfsign_openssl.c [hunk 9]
```diff
@@ -85,27 +86,27 @@ static uint8_t* ReadPkcs12(const char* Name, uint32_t* Length) {
...
+    _In_opt_z_ EVP_PKEY *CaPKey,
     _In_z_ char *SNI,
-    _In_opt_z_ char *Password,
-    _In_ BOOLEAN OutputPkcs12
+    _In_z_ BOOLEAN GenCaCert,
```
Comment (anrossi) [inline src/platform/selfsign_openssl.c:98]: This isn't null-terminated either, so `_In_` is all that's needed here.

File: src/platform/selfsign_openssl.c (hunk 8 above)
Comment (anrossi) [reply src/platform/selfsign_openssl.c:95]: Also, consider renaming this to `SignerCert` or similar to state the purpose more clearly.

Comment (jbevemyr) [reply src/platform/selfsign_openssl.c:95]: Sorry, I was guessing what the _In_*_ annotations meant. Should have looked it up.

File: src/platform/selfsign_openssl.c [hunk 10]
```diff
@@ -588,27 +808,55 @@
...
-        KeyFileName = QuicTestServerPrivateKeyFilename;
+            CertFileName = QuicTestServerCertFilenameSS;
+            KeyFileName = QuicTestServerPrivateKeyFilenameSS;
 #endif
+        }
     }
 
     CXPLAT_CREDENTIAL_CONFIG_INTERNAL* Params =
-        malloc(sizeof(CXPLAT_CREDENTIAL_CONFIG_INTERNAL) + sizeof(TEMP_DIR_TEMPLATE));
+        malloc(sizeof(CXPLAT_CREDENTIAL_CONFIG_INTERNAL) +
```
[the remark this answers is not in the document]
Comment (jbevemyr) [reply src/platform/selfsign_openssl.c:858]: This is a false failure, right? Or does it require me to change anything? I added a linebreak and changed the size of CXPLAT_CREDENTIAL_CONFIG_INTERNAL by adding another file path array, in line with the others in the same struct.

File: src/platform/selfsign_openssl.c [hunk 11]
```diff
@@ -248,6 +294,93 @@
...
+    }
+
+    Status = CxPlatTlsGenerateSignedCert(
+        CaX509, CaPKey, SNI, FALSE, &X509, &PKey);
+
+    if (Status != QUIC_STATUS_SUCCESS)
+        goto Exit;
+
+    if (CaFileName != NULL) {
+        Fd = fopen(CaFileName, "wb");
```
[the remark this answers is not in the document]
Comment (jbevemyr) [reply src/platform/selfsign_openssl.c:360]: This is also a false failure, right? Or at least one that can be ignored. The file is created in the same way the other files are created. It is just test code after all. Should be ok to create with 0666 mode.

**Comment by nibanks** [comment]
@anrossi could you try to review this again today? Thanks!

File: src/cs/lib/msquic_generated.cs [hunk 12]
```diff
@@ -280,6 +281,9 @@ internal unsafe partial struct QUIC_CREDENTIAL_CONFIG
 
         internal QUIC_ALLOWED_CIPHER_SUITE_FLAGS AllowedCipherSuites;
 
+        [NativeTypeName("const char *")]
```
Comment (anrossi) [inline src/cs/lib/msquic_generated.cs:284]: @nibanks How does .NET handle adding a member in the middle of an object?

Comment (nibanks) [reply src/cs/lib/msquic_generated.cs:284]: It's at the end.

File: src/platform/unittest/TlsTest.cpp [hunk 13]
```diff
@@ -165,6 +206,21 @@ struct TlsTest : public ::testing::TestWithParam<bool>
...
         ClientCertParams = (QUIC_CREDENTIAL_CONFIG*)CxPlatGetSelfSignedCert(CXPLAT_SELF_SIGN_CERT_USER, TRUE);
         ASSERT_NE(nullptr, ClientCertParams);
+
+        CaSelfSignedCertParams = (QUIC_CREDENTIAL_CONFIG*)CxPlatGetSelfSignedCert(CXPLAT_SELF_SIGN_CA_CERT_USER, FALSE);
+        ASSERT_NE(nullptr, CaSelfSignedCertParams);
+        CaSelfSignedCertParamsFlags = CaSelfSignedCertParams->Flags;
+        CaClientCertParams = (QUIC_CREDENTIAL_CONFIG*)CxPlatGetSelfSignedCert(CXPLAT_SELF_SIGN_CA_CERT_USER, TRUE);
+        ASSERT_NE(nullptr, ClientCertParams);
+
+        // We need to switch the ca roots around since the server must
```
Comment (anrossi) [inline src/platform/unittest/TlsTest.cpp:216]: Why the swap instead of just assigning the correct certificate to the right *certparams variable?

File: src/cs/lib/msquic_generated.cs (hunk 12 above)
Comment (anrossi) [reply src/cs/lib/msquic_generated.cs:284]: There's all the pointers to the certificate data after it, I don't think this is the end.

File: src/platform/selfsign_openssl.c (hunk 11 above)
[the remark this answers is not in the document]
Comment (anrossi) [reply src/platform/selfsign_openssl.c:378]: Yeah, this failure is not unique to your change.

File: src/cs/lib/msquic_generated.cs (hunk 12 above)
Comment (nibanks) [reply src/cs/lib/msquic_generated.cs:284]: Look at msquic.h for the actual structure. Everything below this is a property (i.e. just a helper function).

**Review by anrossi — COMMENTED** [review]
Looking good! Just a couple things more and could you add the new flag to the `PlatformSpecificFlagsOpenSsl` test in TlsTest.cpp?
Thank you for your patience and effort

File: src/platform/unittest/TlsTest.cpp (hunk 13 above)
Comment (jbevemyr) [reply src/platform/unittest/TlsTest.cpp:216]: It is tricky for two reasons, the certparams are returned by the CxPlatGetSelfSignedCert, so in order to set them directly the ca certificate file parameter would have to be passed as an argument to CxPlatGetSelfSignedCert. However, the ca certificate file is also created by the CxPlatGetSelfSignedCert so we do not even know what to pass in until at least the "other" version returns. The only solution I see, which I've also considered for other reasons, would be to have a new function that creates both the client and server certificates in one go. That way the ca certificates can be assigned correctly from the start. Should I rework the code with this change?

Comment (anrossi) [reply src/platform/unittest/TlsTest.cpp]: Hmm, maybe the CA certificate creation should be split out into its own function which returns the file path to the CA certificate, and then it can be passed as a parameter to `CxPlatGetSelfSignedCert` to sign the client/server certificate?  I think that would be the most flexible design. Does that sound too complicated?

Comment (jbevemyr) [reply src/platform/unittest/TlsTest.cpp]: Sounds like a good plan, I'll give it a try and see what it looks like.

Comment (jbevemyr) [reply src/platform/unittest/TlsTest.cpp]: I pushed a commit with this change, but it introduces quite a bit of code just to avoid the swap...

File: src/platform/selfsign_capi.c [hunk 14]
```diff
@@ -1031,13 +1031,25 @@ FindOrCreateCertificate(
     return (void*) Cert;
 }
 
+_IRQL_requires_max_(PASSIVE_LEVEL)
+const char*
+CxPlatGetSelfSignedCertCaCertificateFileName(
+    _In_ BOOLEAN ClientCertificate
+    )
+{
```
Comment (anrossi) [inline src/platform/selfsign_capi.c:1039]: you'll also need a `UNREFERENCED_PARAMETER(ClientCertificate)` here, otherwise the compiler will yell.

Comment (jbevemyr) [reply src/platform/selfsign_capi.c:1039]: Right, missed that one.

File: src/inc/msquic.h [hunk 15]
```diff
@@ -319,6 +320,7 @@ typedef struct QUIC_CREDENTIAL_CONFIG {
     void* Reserved; // Currently unused
     QUIC_CREDENTIAL_LOAD_COMPLETE_HANDLER AsyncHandler; // Optional
     QUIC_ALLOWED_CIPHER_SUITE_FLAGS AllowedCipherSuites;// Optional
+    const char* CaCertificateFile;
```
Comment (nibanks) [inline src/inc/msquic.h:323]: Two things:

1. This should have a comment indicating it's optional.
2. You can actually leverage the `Reserved` space instead of appending.

File: src/platform/tls_schannel.c [hunk 16]
```diff
@@ -1006,6 +1006,10 @@ CxPlatTlsSecConfigCreate(
         return QUIC_STATUS_INVALID_PARAMETER;
     }
 
+    if (CredConfig->Flags & QUIC_CREDENTIAL_FLAG_SET_CA_CERTIFICATE_FILE) {
+        return  QUIC_STATUS_NOT_SUPPORTED;
```
Comment (nibanks) [inline src/platform/tls_schannel.c:1010]:
```suggestion
        return QUIC_STATUS_NOT_SUPPORTED;
```

File: src/platform/unittest/TlsTest.cpp [hunk 17]
```diff
@@ -100,11 +105,49 @@ struct TlsTest : public ::testing::TestWithParam<bool>
...
+                nullptr,
+                nullptr,
+                QUIC_ALLOWED_CIPHER_SUITE_NONE,
+                CaClientCertParams->CaCertificateFile
+            };
             CredConfig.Flags |= CredFlags;
             CredConfig.AllowedCipherSuites = CipherFlags;
             Load(&CredConfig, TlsFlags);
         }
+
```
Comment (nibanks) [inline src/platform/unittest/TlsTest.cpp:150]: nit: unnecessary change

File: src/platform/unittest/TlsTest.cpp [hunk 18]
```diff
@@ -1326,6 +1390,37 @@ TEST_F(TlsTest, DeferredCertificateValidationAllow)
...
+        ASSERT_NE(nullptr, ServerContext.State.WriteKeys[QUIC_PACKET_KEY_1_RTT]);
+
+        Result = ClientContext.ProcessData(&ServerContext.State, DefaultFragmentSize, true);
+        ASSERT_TRUE(ClientContext.ReceivedPeerCertificate);
+        ASSERT_TRUE(Result & CXPLAT_TLS_RESULT_HANDSHAKE_COMPLETE);
+    }
+}
+
+#endif
+
```
Comment (nibanks) [inline src/platform/unittest/TlsTest.cpp:1422]: nit: remove extra newline

File: src/platform/unittest/TlsTest.cpp [hunk 19]
```diff
@@ -1835,6 +1930,25 @@ TEST_F(TlsTest, ClientCertificateDeferValidation)
     DoHandshake(ServerContext, ClientContext);
 }
 
+#ifndef _WIN32
```
Comment (nibanks) [inline src/platform/unittest/TlsTest.cpp:1933]: Generally, we want more specific flags such as the `QUIC_ENABLE_ANON_CLIENT_AUTH_TESTS` below. Not just using `_WIN32`.

File: src/platform/unittest/TlsTest.cpp [hunk 20]
```diff
@@ -1930,8 +1926,7 @@ TEST_F(TlsTest, ClientCertificateDeferValidation)
     DoHandshake(ServerContext, ClientContext);
 }
 
-#ifndef _WIN32
-
+#ifdef QUIC_ENABLE_CA_CERTIFICATE_FILE
```
Comment (nibanks) [inline src/platform/unittest/TlsTest.cpp:1929]: But you still need to define this, right?

**Comment by nibanks** [comment]
Can you please update `docs\api\QUIC_CREDENTIAL_CONFIG.md` too?

File: src/platform/unittest/TlsTest.cpp (hunk 20 above)
Comment (nibanks) [reply src/platform/unittest/TlsTest.cpp:1929]: And the define should specifically specify `_TESTS`

Comment (jbevemyr) [reply src/platform/unittest/TlsTest.cpp:1929]: Yes, good points.

**Review by anrossi — APPROVED** [review]

**Comment by jbevemyr** [comment]
Ready to merge?

**Comment by nibanks** [comment]
Thanks for working on this!!!
